pkg/swarm/swarm.go

```diff
@@ -161,9 +161,32 @@ type Chunk interface {
 	Equal(Chunk) bool
 }
 
+// BatchID type represents unique batch identifier.
+type BatchID []byte
+
+// String implements fmt.Stringer.
+func (b BatchID) String() string {
+	return hex.EncodeToString(b.Bytes())
+}
+
+// Bytes returns raw []bytes value for this BatchID.
+func (b BatchID) Bytes() []byte {
+	return []byte(b)
+}
+
+// HexToBatchID returns BatchID form supplied hex string.
+func HexToBatchID(s string) (BatchID, error) {
+	b, err := hex.DecodeString(s)
+	if err != nil {
+		return nil, err
+	}
+
+	return BatchID(b), nil
+}
+
 // Stamp interface for postage.Stamp to avoid circular dependency
 type Stamp interface {
-	BatchID() []byte
+	BatchID() BatchID
```

Comment:
I dont see why this is required. This additional type would impose some additional code for no reason. I dont believe we need to define this type at all.

BatchID is not a new type. It is the same hash that we use everywhere. So I dont see why we would need a new type for this. Going to string and back is simple one-liner with hex package

Reply:
Defining domain specific types gives you type safety, enables better abstraction, encapsulation and makes code more readable. 

This kind of mechanism was added to language for vary this reason. 

----

This type does impose few lines of additional code. But at the same time it removes a lot of boilerplate code where developers need to format it when writing to logs.

Reply:
> Defining domain specific types gives you type safety, enables better abstraction, encapsulation and makes code more readable. 
> 
> This kind of mechanism was added to language for vary this reason. 
> 

So I do think there could be some use of it. But in this specific instance I don't think this gives us anything additional. 

The only use for it is to print in string format and hex.EncodeToString is already a one-liner and simple way to understand.

Adding more types increases the size of binary. Having dependencies for types across packages is also not favourable. I don't see any rewards in this case. When I look at some new BatchID type, someone might think this is some new construct, which it is not.

> ----
> 
> This type does impose few lines of additional code. But at the same time it removes a lot of boilerplate code where developers need to format it when writing to logs.

The actual change in `cmd/start.go` is the same. I don't see any boiler plate code saved?

Reply:
BatchID is not a new type. It is essentially a swarm.Address in byte format. So adding a new type to swarm package is definitely not the right thing to do.

Reply:
Alright. In that case `BatchID` could be type alias for `swarm.Address`. 

Although I would add that type alias later since currently `swarm.Address` is unnecessary boxed in a structure which would introduce more changes.

Reply:
> Adding more types increases the size of binary. 
Please, provide me with info on exact increase of the binary in with this type. Also please calculate reduction of the binary where boilerplate calls for `hex.EncodeToString` function is not removed. Lastly we would need to consider current binary size to know impact.

> Having dependencies for types across packages is also not favourable. 
There is no problem importing right types at right places.

>  I don't see any rewards in this case
1) type safety - it's preventing developers to mix frogs with apples
2) provides additional functionality - in this case removes boilerplate code where developer needs to call `hex.EncodeToString` each time when this type is written to the logs
3) makes code easier to read - consider case when you have `b []bytes` and `b BatchID` what is more readable?

> When I look at some new BatchID type, someone might think this is some new construct, which it is not.
It's vary common practice in go to have types for such cases.

Reply:
> The actual change in cmd/start.go is the same. I don't see any boiler plate code saved? 

`cmd/start.go` is bootstrap code, right? 

Boilerplate code is any code that gets repeated in various places with little to no variation.

Reply:
> > Adding more types increases the size of binary. 
> 
> Please, provide me with info on exact increase of the binary in with this type. Also please calculate reduction of the binary where boilerplate calls for `hex.EncodeToString` function is removed. Lastly we would need to consider current binary size to know impact.

I am not sure if you understand what I mean.  Boilerplate code doesnt decrease binary size as such. Compiler anyway would optimize these things. Aliases add more symbols to the codebase. In any case this may not be significant.
> 
> > Having dependencies for types across packages is also not favourable. 
> 
> There is no problem importing right types at right places.
> 
> >  I don't see any rewards in this case
> 
> 1) type safety - it's preventing developers to mix frogs with apples
> 2) provides additional functionality - in this case removes boilerplate code where developer needs to call `hex.EncodeToString` each time when this type is written to the logs
> 3) makes code easier to read and adds more semantics - consider case when you have `b []bytes` and `b BatchID`, later code is much clearer to the developer then first.
> 
> > When I look at some new BatchID type, someone might think this is some new construct, which it is not.
> 
> It's vary common practice in go to have types for such cases. 
> 

It is very common practice to not do unnecessary refactorings on large codebases.

The only reason you were able to make these changes is, the code is already simple enough for anyone to read and understand in the parts that you have touched. As such the changes to add more types adds complexity as developers have to know about this additional type which has no advantage. []byte slice is already used widely and has strong stdlib support for any operations.

Secondly there is not benefit of your changes. Hex.EncodeToString is a one-liner. Anyone who has worked in golang would know about this function and how to use it.

Secondly the boilerplate code you mentioned is actually doing the same thing. The Decode function has no benefit as it doesn't save any lines of code. Developer just has to know about this additional function when he can and knows how to use the hex.Decode function.

Finally, I mentioned in your last PR as well. The correct process to do refactorings is to first create an issue in the bee repo to explain what you are planning to do and get comments from the team. We are the ones who have to maintain the code and solve the bugs. So such large refactorings without prior discussions should be avoided.

Reply:
> It is very common practice to not do unnecessary refactorings on large codebases.

Code is either refactored or new project is started from scratch. Personally, I wouldn't consider this as large codebase.

> Secondly there is not benefit of your changes. Hex.EncodeToString is a one-liner. Anyone who has worked in golang would know about this function and how to use it.

With this change it is not necessary to to even do that (type does it for you 🙂). If we look at the root issue we could see that developers actually used `fmt.Sprintf` instead of `hex.EncodToString`. Certainly there are benefit to this (already mentioned).